src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+// Intersect both vertex sets: these are the candidate vertices.
+// For all candidate vertices c a (potentially arbitrarily bad) alternative route is (s, c, t).
+// Apply heuristic to evaluate alternative route based on stretch, overlap, how reasonable it is.
+//
+// For MLD specifically we can pull off some tricks to make evaluating alternatives fast:
+//   Only consider (s, c, t) with c border vertex: re-use MLD search steps.
+//   Add meta data to border vertices: consider (s, c, t) only when c is e.g. on a highway.
+//   Prune based on vertex cell id
+//
+// https://github.com/Project-OSRM/osrm-backend/issues/3905
+InternalManyRoutesResult
+alternativePathSearch(SearchEngineData<Algorithm> &search_engine_data,
+                      const datafacade::ContiguousInternalMemoryDataFacade<Algorithm> &facade,
+                      const PhantomNodes &phantom_node_pair)
+{
+    const Partition &partition = facade.GetMultiLevelPartition();
+
+    search_engine_data.InitializeOrClearFirstThreadLocalStorage(facade.GetNumberOfNodes());
+
+    Heap &forward_heap = *search_engine_data.forward_heap_1;
+    Heap &reverse_heap = *search_engine_data.reverse_heap_1;
+
+    insertNodesInHeaps(forward_heap, reverse_heap, phantom_node_pair);
+
+    // Saves nodes in the forward and backward search space overlap as candidates
+
+    // The single via node in the shortest paths s,via and via,t sub-paths and
+    // the weight for the shortest path s,t we return and compare alternatives to.
+    NodeID shortest_path_via = SPECIAL_NODEID;
+    EdgeWeight shortest_path_weight = INVALID_EDGE_WEIGHT;
+
+    // The current via node during search spaces overlap stepping and an artificial
+    // weight (overlap factor * shortest path weight) we use as a termination criteria.
+    NodeID overlap_via = SPECIAL_NODEID;
+    EdgeWeight overlap_weight = INVALID_EDGE_WEIGHT;
+
+    // All via nodes in the overlapping search space (except the shortest path via node).
+    // Will be filtered and ranked and then used for s,via and via,t alternative paths.
+    std::vector<WeightedViaNode> candidate_vias;
+
+    // The logic below is a bit weird - here's why: we want to re-use the MLD routingStep for
+    // stepping our search space from s and from t. We don't know how far to overlap until we have
+    // the shortest path. Once we have the shortest path we can use its weight to terminate when
+    // we're over factor * weight. We have to set the weight for routingStep to INVALID_EDGE_WEIGHT
+    // so that stepping will continue even after we reached the shortest path upper bound.
+
+    const bool force_loop_forward = needsLoopForward(phantom_node_pair);
+    const bool force_loop_backward = needsLoopBackwards(phantom_node_pair);
+
+    EdgeWeight forward_heap_min = forward_heap.MinKey();
+    EdgeWeight reverse_heap_min = reverse_heap.MinKey();
+
+    while (forward_heap.Size() + reverse_heap.Size() > 0)
+    {
+        if (shortest_path_weight != INVALID_EDGE_WEIGHT)
```

Comment:
Do you think we could make this process less brittle by not assigning/updating/re-using weights? Overlap weight is assigned here for a single check, before it is cleared to Invalid, updated in the routing step and again invalidated. This can get somewhat confusing.